brainiak/reprsimil/brsa.py

```diff
@@ -0,0 +1,1316 @@
+
+        return current_vec_U_chlsk_l_AR1, current_a1, current_logSNR2,\
+            current_GP
+
+    # We fit two parts of the parameters iteratively.
+    # The following are the corresponding negative log likelihood functions.
+
+    def _loglike_y_AR1_diagV_fitU(self, param, XTX, XTDX, XTFX, YTY_diag,
+                                  YTDY_diag, YTFY_diag, XTY, XTDY, XTFY,
+                                  log_SNR2, l_idx, n_C, n_T, n_V, rank):
+        # This code calculates the log likelihood of data given cholesky
+        # decomposition of U and AR(1) parameters of noise as free parameters.
+        # Free parameters are in param.
+        # The log of the square of signal to noise level in each voxel
+        # (the ratio of the diagonal elements in V and
+        # the noise variance) are fixed. This likelihood is iteratively
+        # optimized with the one with suffix _fitV.
+        #
+        # The meaing of U and V follow this wiki page of matrix normal
+        # distribution:
+        # https://en.wikipedia.org/wiki/Matrix_normal_distribution
+        #
+        # We assume betas of all voxels as a matrix follow this distribution.
+        # U describe the covariance between conditions. V describe the
+        # covariance between voxels.
+        #
+        # In this version, we assume that beta is independent between voxels
+        # and noise is also independent.
+        # By the assumption that noise is independent, we only need to pass
+        # the products X'X, X'Y and Y'Y, instead of X and Y
+        # Y'Y is passed in the form of its diagonal elements.
+        # DiagV means we assume that the variance of beta can be different
+        # between voxels. This means that V is a diagonal matrix instead of
+        # an identity matrix. The parameter includes the lower triangular
+        # part of the cholesky decomposition
+        # of U (flattened), then tan(rho1*pi/2) where rho1 is
+        # each voxel's autoregressive coefficient (assumging AR(1) model).
+        # Such parametrization avoids the need of boundaries
+        # for parameters.
+
+        LL = 0.0  # log likelihood
+
+        n_l = np.size(l_idx[0])
+        # the number of parameters in the index of lower-triangular matrix
+        # This indexing allows for parametrizing only
+        # part of the lower triangular matrix (non-full rank covariance matrix)
+        L = np.zeros([n_C, rank])
+        # lower triagular matrix L, cholesky decomposition of U
+        L[l_idx] = param[0:n_l]
+
+        a1 = param[n_l:n_l + n_V]
+        rho1 = 2.0 / np.pi * np.arctan(a1)  # auto-regressive coefficients
+
+        SNR2 = np.exp(log_SNR2)
+        # each element of SNR2 is the ratio of the diagonal element on V
+        # to the variance of the fresh noise in that voxel
+
+        # derivatives
+        deriv_L = np.zeros(np.shape(L))
+        deriv_a1 = np.zeros(np.shape(rho1))
+
+        YTAY = YTY_diag - rho1 * YTDY_diag + rho1**2 * YTFY_diag
```

Comment:
Would be really really good to reference these lines to expressions / subexpressions in the paper.